wingui: Move window procedure to separate thread to fix hanging/ugliness during resize/move

loblolly986 opened the pull request:
Following up on issue #156, here's a proper way to fix the application hanging during a window resize or move.

I'm aware that WinGUI moved away from some sort of multi-threaded implementation several years back, but the synchronization code needed in this case is really pretty minimal and non-invasive. It uses a "critical section" object for synchronization. 

Before the window thread executes code that accesses or modifies sensitive shared variables, it requests ownership of the critical section object. The main thread owns the critical section except when it calls `PDC_napms()` or needs to make a Win32 API call that would *send* (as distinguished from "post") a message to the window and wait for it to be processed (the `SetWindowText()` call in `PDC_set_title()` is one instance, as is `SetWindowPos()` in `PDC_resize_screen()`). Before returning, the window procedure "leaves" the critical section. From my testing, the stability of this implementation seems pretty solid.

Features:

- Synchronization calls for the window thread are contained entirely in the window procedure, and only a few locations used by the main thread needed them.
- Low CPU footprint when idle. Because the main thread no longer has to handle the message queue, I was able to simplify `PDC_napms()` down to just a `Sleep()` call wrapped with `LeaveCriticalSection()`/`EnterCriticalSection()`. The window thread uses `GetMessage()` in its message loop, which blocks until a message is received.
- I removed or commented out up some superfluous code while I was at it, such as the `PDC_napms()` call in `PDC_check_key()` (the removal of which also improves the accuracy of `halfdelay()` - more on that later) and the `wndclass_has_been_registered` stuff in `set_up_window()` (more than one window is not currently supported). `PDC_bDone` is also no longer needed, and active instances of it have been commented out.

Some notes:

- The responsiveness of an application to resize events lags slightly. This can be improved by reducing the 50 ms checking interval in `wgetch()` to something like 1 or 5 ms. But this would affect the other PDCurses ports as well (unless wrapped with `#ifdef`s) and mess up `halfdelay()` on some systems (the precision of `Sleep()` is limited depending on the OS and/or hardware; `Sleep(1)` or `Sleep(5)` can result in a sleep of 15 ms or more).
- Bill-Gray's tweak in commit 0a511bfae15f4aa36917f37525676accd741f68f isn't needed anymore, but I kept it because combining multiple `KEY_RESIZE` events is a good idea that improves efficiency with applications with relatively "expensive" redraws.
- My code uses `_beginthreadex()` to create the window thread to ensure that C runtime library functions can be safely used. This should be fine with Visual C++, MinGW, MinGW-w64, Open Watcom, and later Borland C++ versions (I have tested with 32-bit MinGW-w64). But it may cause compatibility issues with early Borland C++ versions; my understanding is that Borland only offered `_beginthread()` and their own `_beginthreadNT()` to start with, then defined `_beginthreadex()` as a macro that resolves to `_beginthreadNT()` (around the time of Borland C++ 5.0), before finally properly implementing `_beginthreadex()`. My code waits on the handle returned by the function to catch premature thread termination, which makes using `_beginthread()` (or Borland's `_beginthreadNT()`) unsafe because it automatically closes the handle when the thread exits. Not sure about old versions of Digital Mars. I could probably implement this another way if supporting those earlier commercial compiler versions is still desired, but it would be less efficient.

Bill-Gray commented:
Looks as if you're on to something good here.  Just a few comments :

- I have Borland C++ 5.5,  and `beginthreadex()` is a separate function from `beginthreadNT()`.  Digital Mars appears to support it properly as well.  I can't come up with a good reason to support older versions of either compiler.  So I think we're good on that point.
- I did remove the multi-threadedness from WinGUI some years back.  But that was because I realized there was a straightforward way to do so and that the multi-threadedness had been rather foolish of me.  I think we're stuck with it here,  and it looks to me as if you've got it implemented about as cleanly as it can be.
- I'll give this a go on my usual test platforms (Wine,  Win10,  and WinME) and then likely pull it.  There's some chance I won't have time to do all of it;  if I don't get around to it in the next few days, I'll just trust you've got it right and pull it in.  I gather you've tested Win10,  so I really only need look at the other two anyway.

loblolly986 commented:
My test systems run Windows NT 4.0 (with SP6a) and 32-bit Vista, both with dual-core processors. I don't have and have no intention to get Windows 10. Though for what it's worth, while I primarily consult the Visual Studio 97 MSDN Library (included with Visual C++ 5.0), which details the (documented) APIs for 95, NT 4.0, and Win32s (useful since the modern docs invariably contain bald-faced lies about the minimum-supported OS version), I also consult the modern online Microsoft documentation as sometimes it will contain improvements, or additional information pertaining to more modern versions.

I considered bringing this up above, but I've noticed that while the window procedure properly returns after processing some messages, with others (including messages it doesn't handle), it breaks from the switch statement, calls `add_mouse()` (regardless of whether a mouse event was encountered), and then calls `DefWindowProc()`. The Microsoft documentation for all of these other messages says we should return 0 for events we processed, and having to call `add_mouse()` apart from mouse and `WM_TIMER` messages seems superfluous and prevents the window procedure from executing asynchronously most of the time even with messages we don't handle, thus reducing efficiency.

I've gone ahead and tried changing it so that it only uses the mouse code at the end of the procedure following mouse button/move and `WM_TIMER` messages, enters the critical section only when needed for the given message, and only calls `DefWindowProc()` when appropriate. As far as I can tell from testing with e.g. NewTest and TestCurs, the mouse behavior seems the same with this revision as it was before, so I've gone ahead and added it to this pull request.

As a bonus, this revision also modifies `PDC_choose_a_new_font()` so the application (main thread) is no longer hung while the font dialog is displayed. This now makes one instance of window thread synchronization calls outside the window procedure, but I think it's worth it. I also fixed a bug where the size selected in the font dialog was being saved even if the user clicked the "Cancel" button.

Bill-Gray commented:
> My test systems run Windows NT 4.0 (with SP6a) and 32-bit Vista, both with dual-core processors. I don't have and have no intention to get Windows 10

You're not missing much.  (More seriously,  it sounds as if we're really testing four separate forks of Windows between us,  plus Wine.  Which is probably a good way to catch bugs.)

Hadn't noticed the app hanging while in the font dialog,  but it does do that under Wine (and ceases to do so with your patch applied.)  I can run `worm.exe` and resize the window,  and
the worms keep moving.  Compiles and runs with OpenWATCOM 1.9 and cross-compiled with MinGW.

I'm probably missing something,  but... in `WndProc`,  when handling the various button messages,  you've added an `EnterCriticalSection(&PDC_cs);` for each one.  It looks to me as if you could just add one such statement,  further down,  under `/* mouse handling code */`.

As it stands,  it's a little hard to tell if each 'Enter' has a matching 'Leave'.

loblolly986 commented:
Good catch, I don't always notice how things can be simplified further right away; since those mouse button entries only modify variables local to `WndProc()` before `break`ing, and since `add_mouse(-1, -1, -1, -1)` is now only called for two messages, we can just call it there and simplify the other mouse message entries as you suggest. I've added another revision to this PR.

Bill-Gray commented:
Took me a while to check,  but it works nicely on Win10 and WinME and Wine/Xubuntu,  plus your tests... so I'll call it good and pull it.  Thanks!
